### ISSUE TITLE
Reconsider harvest flow architecture

### ISSUE THREAD

**Issue by alexdunnjpl** [NASA-PDS/harvest#114]
## 🐛 Describe the bug
From #112 [comment](https://github.com/NASA-PDS/harvest/issues/112#issuecomment-1341902833)

> The way I'd expect harvest to work is that there would be
> - an enumeration step (enumeration of the products* to be processed)
> - a generation step (generation of the registry documents for each product)
> - a write step (i.e. throwing those documents at the registry database, and handling any errors
> *product/collection/bundle
> 
> this way, different enumeration approaches (in this case, `<bundles>` and `<directories>`) would be decoupled from  everything else, and share a common downstream processing path.
> 
> The current situation is that `<bundle>` leverages `HarvestCmd.processBundles` and `<directories>` leverages `HarvestCmd.processDirectories` (likewise, files and collections have equivalents), and these processing execution paths diverge completely depending on which kind of source they're deriving from.

This _may_ be something we want to consider rearchitecting.  The fact that there are a bunch of different execution paths (depending on how artifacts are targeted) for flows which are ostensibly identical (registering artifacts) is a bit spaghetti-ish, and can result in [changes which are intended to be global (like the implementation of an `--overwrite` CLI option) only affecting some execution paths.](https://github.com/NASA-PDS/harvest/issues/112#issuecomment-1341902833).

The lack of separation of concerns also makes it a bit trickier (I need to investigate) to confidently make [changes which are intended to only affect enumeration behaviour](https://github.com/NASA-PDS/harvest/issues/102).

The flipside of this is that the scope of the refactor is something that requires either familiarity with the whole of `harvest`, or solid tests (preferably both) to perform in the first place, if my gut hunch is correct.

**Comment by tloubrieu-jpl** [comment]
@alexdunnjpl Group this thinking with the scalable harvest components.

**Comment by alexdunnjpl** [comment]
Given that harvest-service and (standalone) harvest are parallel development projects.

Tentative thoughts:

1. Identify infra-agnostic stages of the process, such as 
    1. Given a file-system root node (ie path), enumerate all products (Bundles, Collections, and simple Products) under that node's subtree.
    2. Given an enumeration of products, map them to registry JSON documents.
    3. Given a batch of registry JSON documents, process their registration with an OpenSearch instance (maybe - is this infra-dependent?)
2. Extract each of these stages as utility modules (using standalone harvest for development, since that will be simplest).
3. Once these utility modules are written/extracted, replace harvest-service implementations with calls to the modules as dependencies.

Once complete, the implementation code of harvest-service will just be the management/delegation code and some simple calls to glue it to the utility modules, and the implementation code of (standalone) harvest will just be a CLI wrapper around calls to the utility modules.

As a result, each (standalone/scalable) version of harvest will be doing exactly the same thing, and the utility libraries will be easily unit-testable.

**Comment by alexdunnjpl** [comment]
Per @jordanpadams 

> 
This is true. <bundles> was kept for backwards compatibility support, but we changed the way we treated this when we decided to just use this part of the config to know where to look, but no longer decipher between bundles/collections/products. Just read everything below where this points and load the data as fast as you can.

Currently unclear whether removal of <bundles> support is now acceptable - will be determined in tomorrow's meeting.

**Comment by jordanpadams** [comment]
@alexdunnjpl this ticket is 2 years old. Is this still applicable?

**Comment by alexdunnjpl** [comment]
@al-niessner does harvest still have early-bifurcating behaviour?  I'll defer to you whether my original suggestion is

- still applicable
- worth pursuing

since I haven't touched harvest in ages.

**Comment by al-niessner** [comment]
@alexdunnjpl 

Yes, there are still 3 or  top level paths: bundles, files, and directories. They seem different in the first 10 lines but then combine further down the processing steam resulting in consistent behavior with respect to the database (not loading data from disk).

I would not waste effort in rewriting harvest. It desperately needs to be rewritten, but user expectations are set making it too late. You would spend so much effort making it misbehave the way the users now expect, any rewrite would become convoluted and embarrassing.

**Comment by alexdunnjpl** [comment]
@jordanpadams suggest closing on that basis ^^^^

Thanks Al
